File: src/main/java/dark/leech/text/plugin/vbook/VBookToLeechTextConverter.java
package dark.leech.text.plugin.vbook;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.HexFormat;
import java.util.Map;

import dark.leech.text.enities.PluginEntity;
import dark.leech.text.plugin.vbook.model.VBookPluginEntity;

/**
 * Converts vBook plugin format to LeechText PluginEntity format. Handles metadata mapping, script
 * inlining, and format transformations.
 */
public class VBookToLeechTextConverter {

    /**
     * Convert a vBook plugin to LeechText PluginEntity.
     *
     * @param vbookPlugin The vBook plugin to convert
     * @return PluginEntity in LeechText format
     * @throws IllegalArgumentException if required fields are missing
     */
    public PluginEntity convert(VBookPluginEntity vbookPlugin) {
        if (vbookPlugin == null) {
            throw new IllegalArgumentException("vbookPlugin cannot be null");
        }

        validateRequiredFields(vbookPlugin);

        // Build plugin entity using chained builder methods
        var entity =
                PluginEntity.builder()
                        .name(vbookPlugin.getName())
                        .author(vbookPlugin.getAuthor() != null ? vbookPlugin.getAuthor() : "vBook")
                        .version(convertVersion(vbookPlugin.getVersion()))
                        .source(vbookPlugin.getSource())
                        .regex(vbookPlugin.getRegexp())
                        .describe(vbookPlugin.getDescription())
                        .language(mapLocaleToLanguage(vbookPlugin.getLocale()))
                        .group(mapTypeToGroup(vbookPlugin.getType()))
                        .icon(vbookPlugin.getIconBase64())
                        .supportUpdate(true)
                        .checked(true)
                        .uuid(generateUuid(vbookPlugin.getName(), vbookPlugin.getVersion()));

        // Inline script contents
        Map<String, String> contents = vbookPlugin.getScriptContents();
        if (contents != null) {
            if (contents.containsKey("chap")) {
                entity.chapGetter(contents.get("chap"));
            }
            if (contents.containsKey("toc")) {
                entity.tocGetter(contents.get("toc"));
            }
            if (contents.containsKey("detail")) {
                entity.detailGetter(contents.get("detail"));
            }
            if (contents.containsKey("page")) {
                entity.pageGetter(contents.get("page"));
            }
            if (contents.containsKey("search")) {
                entity.searchGetter(contents.get("search"));
            }

            // Map vBook-specific scripts to appropriate getters or skip
            // home → Not used in LeechText, can add to description
            // genre → Not used in LeechText, can add to description
            // site → Not used in LeechText, can add to description
        }

        return entity.build();
    }

    /** Validate that required fields are present. */
    private void validateRequiredFields(VBookPluginEntity plugin) {
        if (plugin.getName() == null || plugin.getName().isEmpty()) {
            throw new IllegalArgumentException("Plugin name is required");
        }
        if (plugin.getRegexp() == null || plugin.getRegexp().isEmpty()) {
            throw new IllegalArgumentException("Plugin regexp is required");
        }
        Map<String, String> contents = plugin.getScriptContents();
        if (contents == null || contents.isEmpty()) {
            throw new IllegalArgumentException("Plugin must have at least one script");
        }
    }

    /**
     * Convert vBook integer version to LeechText double version. vBook uses integers like 10,
     * LeechText uses doubles like 1.0 Formula: int_version / 10.0
     *
     * @param vbookVersion Integer version from vBook
     * @return Double version for LeechText
     */
    private double convertVersion(Integer vbookVersion) {
        if (vbookVersion == null) {
            return 1.0;
        }
        return vbookVersion / 10.0;
    }

    /**
     * Map vBook locale to LeechText language code. vBook uses full locale (vi_VN), LeechText uses
     * short codes (vi)
     *
     * @param locale Locale from vBook (e.g., "vi_VN")
     * @return Language code for LeechText (e.g., "vi")
     */
    private String mapLocaleToLanguage(String locale) {
        if (locale == null || locale.isEmpty()) {
            return "en";
        }

        // Extract language code from locale (vi_VN → vi)
        String[] parts = locale.split("_");
        String langCode = parts[0].toLowerCase();

        // Map common language codes
        return switch (langCode) {
            case "vi" -> "vi";
            case "en" -> "en";
            case "zh" -> "cn";
            default -> langCode;
        };
    }

    /**
     * Map vBook type to LeechText group. vBook uses "novel" or "comic", LeechText uses group names
     *
     * @param type Type from vBook ("novel", "comic")
     * @return Group name for LeechText
     */
    private String mapTypeToGroup(String type) {
        if (type == null || type.isEmpty()) {
            return "dich"; // default group
        }

        return switch (type.toLowerCase()) {
            case "novel" -> "dich";
            case "comic" -> "truyentranh";
            default -> "dich";
        };
    }

    /**
     * Generate UUID from plugin name and version. Creates a consistent UUID based on SHA-256 hash.
     *
     * @param name Plugin name
     * @param version Plugin version
     * @return UUID string
     */
    private String generateUuid(String name, Integer version) {
        try {
            String input = name + "-" + (version != null ? version : "1");
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));

            // Convert first 16 bytes to UUID format
            HexFormat hex = HexFormat.of();
            StringBuilder sb = new StringBuilder();

            for (int i = 0; i < 16; i++) {
                if (i == 4 || i == 6 || i == 8 || i == 10) {
                    sb.append("-");
                }
                sb.append(hex.toHexDigits(hash[i]));
            }

            return sb.toString();

        } catch (Exception e) {
            // Fallback to simple hash-based UUID
            return String.format(
                    "%08x-%04x-%04x-%04x-%012x",
                    name.hashCode(),
                    version != null ? version : 1,
                    0,
                    0,
                    System.currentTimeMillis() & 0xffffffffffffL);
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/vbook/VBookToLeechTextConverter.kt
package dev.haipham22.leechtext.plugin.vbook

import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.plugin.vbook.model.VBookPluginEntity
import dev.haipham22.leechtext.util.nowMillis
import dev.haipham22.leechtext.util.sha256Hex

/**
 * Converter vBook plugin → LeechText PluginEntity (port từ vbook/VBookToLeechTextConverter.java).
 * Metadata mapping, script inlining, format transformation.
 */
class VBookToLeechTextConverter {
    /**
     * Convert vBook plugin → PluginEntity LeechText format.
     *
     * @throws IllegalArgumentException nếu thiếu required fields
     */
    fun convert(vbookPlugin: VBookPluginEntity?): PluginEntity {
        requireNotNull(vbookPlugin) { "vbookPlugin cannot be null" }

        validateRequiredFields(vbookPlugin)

        val entity =
            PluginEntity(
                name = vbookPlugin.name,
                author = vbookPlugin.author ?: "vBook",
                version = convertVersion(vbookPlugin.version),
                source = vbookPlugin.source,
                regex = vbookPlugin.regexp,
                describe = vbookPlugin.description,
                language = mapLocaleToLanguage(vbookPlugin.locale),
                group = mapTypeToGroup(vbookPlugin.type),
                icon = vbookPlugin.iconBase64,
                supportUpdate = true,
                checked = true,
                uuid = generateUuid(vbookPlugin.name, vbookPlugin.version),
            )

        // Inline script contents (immutable view — converter không đụng map gốc)
        val contents: Map<String, String> = vbookPlugin.scriptContents
        contents["chap"]?.let { entity.chapGetter = it }
        contents["toc"]?.let { entity.tocGetter = it }
        contents["detail"]?.let { entity.detailGetter = it }
        contents["page"]?.let { entity.pageGetter = it }
        contents["gen"]?.let { entity.genGetter = it }
        contents["search"]?.let { entity.searchGetter = it }
        contents["home"]?.let { entity.homeGetter = it }
        contents["genre"]?.let { entity.genreGetter = it }
        contents["tab"]?.let { entity.tabGetter = it }
        contents["gen"]?.let { entity.genGetter = it }

        entity.tag = vbookPlugin.tag
        entity.priority = vbookPlugin.priority ?: 0
        // Toàn bộ src/*.js → load() trong sandbox resolve in-memory (config.js, helper...)
        entity.extraScripts = vbookPlugin.scriptContents.entries.associate { (k, v) -> "$k.js" to v }

        // Config: key → value (sandbox inject) + spec cho UI cấu hình.
        // Built-in keys (vBook extension-api): thread_num, delay, timeout — luôn có defaults
        val configMap = mutableMapOf<String, String>()
        vbookPlugin.config.forEach { (key, spec) ->
            configMap[key] = spec.default ?: ""
        }
        // Built-in connection settings — mọi plugin đều có
        configMap.putIfAbsentCompat("thread_num", "3")
        configMap.putIfAbsentCompat("delay", "0")
        configMap.putIfAbsentCompat("timeout", "30000")
        entity.config = configMap

        // Spec cho UI: declared config + built-in
        val specMap =
            vbookPlugin.config
                .mapValues { (_, spec) ->
                    dev.haipham22.leechtext.entities.PluginConfigSpec(
                        title = spec.title,
                        subtitle = spec.subtitle,
                        default = spec.default,
                        values = spec.values,
                        mode = spec.mode,
                        format = spec.format,
                    )
                }.toMutableMap()
        specMap.putIfAbsentCompat(
            "thread_num",
            dev.haipham22.leechtext.entities.PluginConfigSpec(
                title = "Số luồng tải",
                subtitle = "Max concurrent requests",
                default = "3",
                mode = "input",
                format = "number",
            ),
        )
        specMap.putIfAbsentCompat(
            "delay",
            dev.haipham22.leechtext.entities.PluginConfigSpec(
                title = "Delay giữa request (ms)",
                subtitle = "Delay between requests",
                default = "0",
                mode = "input",
                format = "number",
            ),
        )
        specMap.putIfAbsentCompat(
            "timeout",
            dev.haipham22.leechtext.entities.PluginConfigSpec(
                title = "Timeout request (ms)",
                subtitle = "Request timeout",
                default = "30000",
                mode = "input",
                format = "number",
            ),
        )
        entity.configSpec = specMap

        return entity
    }

    /** Validate required fields. */
    private fun validateRequiredFields(plugin: VBookPluginEntity) {
        require(!plugin.name.isNullOrEmpty()) { "Plugin name is required" }
        require(!plugin.regexp.isNullOrEmpty()) { "Plugin regexp is required" }
        require(plugin.scriptContents.isNotEmpty()) { "Plugin must have at least one script" }
    }

    /**
     * Version integer vBook → double LeechText (10 → 1.0). Formula: int_version / 10.0.
     */
    private fun convertVersion(vbookVersion: Int?): Double = vbookVersion?.div(10.0) ?: 1.0

    /** Locale vBook ("vi_VN") → language code LeechText ("vi"). */
    private fun mapLocaleToLanguage(locale: String?): String {
        if (locale.isNullOrEmpty()) return "en"

        // Extract language code (vi_VN → vi)
        val langCode = locale.split("_")[0].lowercase()

        return when (langCode) {
            "vi" -> "vi"
            "en" -> "en"
            "zh" -> "cn"
            else -> langCode
        }
    }

    /** Type vBook ("novel"/"comic") → group LeechText. */
    private fun mapTypeToGroup(type: String?): String {
        if (type.isNullOrEmpty()) return "dich" // default group

        return when (type.lowercase()) {
            "novel" -> "dich"
            "comic" -> "truyentranh"
            else -> "dich"
        }
    }

    /**
     * UUID nhất quán từ name + version (SHA-256 hash, 16 bytes đầu → format UUID).
     */
    private fun generateUuid(
        name: String?,
        version: Int?,
    ): String = try {
        val input = "$name-${version ?: 1}"
        val hex = sha256Hex(input.encodeToByteArray()).take(32)
        // Convert 16 bytes đầu (32 hex chars) sang UUID format
        val sb = StringBuilder()
        for (i in 0 until 16) {
            if (i in listOf(4, 6, 8, 10)) sb.append("-")
            sb.append(hex.substring(i * 2, i * 2 + 2))
        }
        sb.toString()
    } catch (e: Exception) {
        // Fallback hash-based UUID
        val ts = nowMillis() and 0xffffffffffffL
        val h = (name?.hashCode() ?: 0).toUInt().toString(16).padStart(8, '0')
        val v = (version ?: 1).toString(16).padStart(4, '0')
        "$h-$v-0000-0000-${ts.toString(16).padStart(12, '0')}"
    }
}

/** putIfAbsent common (MutableMap JVM-only). */
private fun <K, V> MutableMap<K, V>.putIfAbsentCompat(
    key: K,
    value: V,
): V? {
    val existing = this[key]
    if (existing == null) this[key] = value
    return existing
}
